cpp/src/arrow/ipc/adapter.cc

```diff
@@ -243,6 +271,26 @@ class RowBatchReader::Impl {
       return MakePrimitiveArray(
           type, field_meta.length, data, field_meta.null_count, null_bitmap, out);
     }
+    if (IsListType(type.get())) {
```

Comment:
I will check what clang-tidy says.  In other cases it liked not having else if when the original "if" block always returned.   If we prefer to allow either I can look into turning off that particular check.